#### step-by-step also supported
Genome.fa.size is required : generated by: perl CRAQ/src/fetch_size.pl Genome.fa > Genome.fa.size  
1. SMS read mapping, filtering and putative LER calling.
```
$ bash src/runLR.sh -g  Genome.fa -z Genome.fa.size -1 SMS_sort.bam 
```
or 
```     
$ bash src/runLR.sh -g  Genome.fa -z Genome.fa.size -1 Pac/ONT.fa.gz -x map-pb -t 10
```

LRout:  
LR_sort.bam	: Filtered SMS alignment file, for view inspection in genome browser.  
LR_sort.bam.bai	: Index of alignment file.  
LR_sort.depth	: SMS mapping coverage.  
LR_clip.coverRate: All output of SMS clipping positions, with columns:chr, position, strand, number of clipped-reads, and total coverage at the position. The strand is just left-clipped(+) or right-clipped(-) to help identify the clipping orientation.  
LR_putative.ER.HR  : Coordinates of putative structral errors or variant breakages. Filtered from LR_clip.coverRate file.  

2. NGS read mapping, filtering and putative SER calling.
```
$ bash src/runSR.sh -g  Genome.fa -z Genome.fa.size  -1 NGS_sort.bam
```
or
```
$ bash src/runSR.sh -g  Genome.fa -z Genome.fa.size  -1 NGS_R1.fa.gz -2 NGS_R2.fa.gz -t 10
```
SRout:  
SR_sort.bam     : Filtered NGS alignment file, for view inspection in genome browser.  
SR_sort.bam.bai : Index of alignment file.  
SR_sort.depth   : NGS mapping coverage.  
SR_clip.coverRate: All output of NGS clipping positions, with columns:chr, position, strand, number of clipped-reads, and total coverage at that position. The strand is just left-clipped(+) or right-clipped(-) to help identify the clipping orientation.  
SR_putative.ER.HR	: Coordinates of putative small-scale errors or heterozygous indel breakages. Filtered from SR_clip.coverRate file.

Note:  
If user used 'bowtie2' generate shortRead alignment in advance, the '--local'(local alignment) option should be performed for generating clipping signal.  

3. Benchmark genomic quality using AQI.       
```
$ bash src/runAQI.sh -g  Genome.fa -z  Genome.fa.size -e LRout/LR_eff.size  -c SRout/SR_putative.ER.HR -C LRout/LR_putative.ER.HR  -d SRout/SR_sort.depth  -D LRout/LR_sort.depth
``` 
### Main output (runAQI_out):  

locER_out/out_final.SER.bed	: Exact coordinates of small regional errors.  
locER_out/out_final.SHR.bed     : Exact coordinates of small heterozygous indels.  
strER_out/out_final.LER.bed	: Exact coordinates of large structral error breakage.  
strER_out/out_final.LHR.bed	: Exact coordinates of structral heterozygous variants.  
out_regional.Report : Statistics for regional genomic metrics.  
out_final.Report : Summary reports inclinding classfied quality metrics(S-AQI, L-AQI) for single scaffold and whole-assembly.  
out_circos.pdf : Drawing genomic metrics.  
out_correct.fa	: A CRAQ-corrected FASTA fragments generated (if --break|-b T)  

